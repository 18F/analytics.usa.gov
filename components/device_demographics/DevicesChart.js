import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import barChart from "../../js/lib/barchart";
import formatters from "../../js/lib/formatters";
import renderBlock from "../../js/lib/renderblock";
import transformers from "../../js/lib/transformers";

function DevicesChart({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/devices.json`;
  const ref = useRef(null);

  useEffect(() => {
    const initDevicesChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock
            .loadAndRender()
            .transform((d) => {
              const devices = transformers.listify(d.totals.by_device);
              devices.forEach((device) => {
                if (device.key === "smart tv") {
                  device.key = "Smart TV";
                }
              });
              return transformers.findProportionsOfMetricFromValue(devices);
            })
            .render(
              barChart()
                .value((d) => d.proportion)
                .format(formatters.floatToPercent),
            ),
        );
      return result;
    };
    initDevicesChart().catch(console.error);
  }, []);

  return (
    <figure id="chart_device_types" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

DevicesChart.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default DevicesChart;
