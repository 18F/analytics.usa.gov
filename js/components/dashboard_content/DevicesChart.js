import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";
import barChart from "../../lib/chart_helpers/barchart";
import formatters from "../../lib/chart_helpers/formatters";
import transformers from "../../lib/chart_helpers/transformers";

/**
 * Retrieves the devices report from the passed data URL and creates a
 * visualization for the breakdown of devices of users visiting sites for the
 * current agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function DevicesChart({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/devices.json`;
  const ref = useRef(null);
  const [deviceData, setDeviceData] = useState(null);

  useEffect(() => {
    const initDevicesChart = async () => {
      if (!deviceData) {
        const data = await DataLoader.loadJSON(dataURL);
        await setDeviceData(data);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder
          .setElement(ref.current)
          .setData(deviceData)
          .setTransformer((d) => {
            const devices = transformers.listify(d.totals.by_device);
            devices.forEach((device) => {
              if (device.key === "smart tv") {
                device.key = "Smart TV";
              }
            });
            return transformers.findProportionsOfMetricFromValue(devices);
          })
          .setRenderer(
            barChart()
              .value((d) => d.proportion)
              .format(formatters.floatToPercent),
          )
          .build();
      }
    };
    initDevicesChart().catch(console.error);
  }, [deviceData]);

  return (
    <figure id="chart_device_types" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

DevicesChart.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default DevicesChart;
