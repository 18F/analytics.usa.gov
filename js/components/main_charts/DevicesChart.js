import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import barChart from "../../lib/chart_helpers/barchart";
import formatters from "../../lib/chart_helpers/formatters";
import renderBlock from "../../lib/chart_helpers/renderblock";
import transformers from "../../lib/chart_helpers/transformers";

/**
 * Retrieves the devices report from the passed data URL and creates a
 * visualization for the breakdown of devices of users visiting sites for the
 * current agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function DevicesChart({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/devices.json`;
  const ref = useRef(null);

  useEffect(() => {
    const initDevicesChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock
            .loadAndRender()
            .transform((d) => {
              const devices = transformers.listify(d.totals.by_device);
              devices.forEach((device) => {
                if (device.key === "smart tv") {
                  device.key = "Smart TV";
                }
              });
              return transformers.findProportionsOfMetricFromValue(devices);
            })
            .render(
              barChart()
                .value((d) => d.proportion)
                .format(formatters.floatToPercent),
            ),
        );
      return result;
    };
    initDevicesChart().catch(console.error);
  }, []);

  return (
    <figure id="chart_device_types" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

DevicesChart.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default DevicesChart;
