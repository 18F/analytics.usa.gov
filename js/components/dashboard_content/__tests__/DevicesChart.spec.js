import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { delay } from "../../../../spec/support/test_utilities";
import DataLoader from "../../../lib/data_loader";
import DevicesChart from "../DevicesChart";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadJSON: jest.fn(),
}));

describe("DevicesChart", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <DevicesChart dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "devices",
        query: {},
        meta: {
          name: "Devices",
          description: "30 days of desktop/mobile/tablet visits for all sites.",
        },
        totals: {
          visits: 1492377457,
          by_device: {
            mobile: 811958807,
            desktop: 656318825,
            "(other)": 1938220,
            tablet: 19920488,
            "smart tv": 2241117,
          },
        },
        taken_at: "2024-03-11T13:59:08.677Z",
      };

      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <DevicesChart dataHrefBase="http://www.example.com/data/" />,
      );
      await waitFor(() => screen.getByText("mobile"));
      // Wait for barchart transition animation to complete (200 ms, set in
      // js/lib/chart_helpers/barchart.js)
      await delay(300);
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <DevicesChart dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
