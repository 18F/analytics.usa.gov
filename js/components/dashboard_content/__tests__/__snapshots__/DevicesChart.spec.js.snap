// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevicesChart when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <figure
    id="chart_device_types"
  >
    <div
      class="data bar-chart"
    >
      <div
        class="bin"
      >
        <div
          class="label"
        >
          mobile
        </div>
        <div
          class="value"
        >
          54.4%
        </div>
        <div
          class="bar"
          style="width: 54.4%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          desktop
        </div>
        <div
          class="value"
        >
          44%
        </div>
        <div
          class="bar"
          style="width: 44%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          tablet
        </div>
        <div
          class="value"
        >
          1.3%
        </div>
        <div
          class="bar"
          style="width: 1.3%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Smart TV
        </div>
        <div
          class="value"
        >
          0.2%
        </div>
        <div
          class="bar"
          style="width: 0.2%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          (other)
        </div>
        <div
          class="value"
        >
          0.1%
        </div>
        <div
          class="bar"
          style="width: 0.1%;"
        />
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`DevicesChart when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <figure
    id="chart_device_types"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;

exports[`DevicesChart when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <figure
    id="chart_device_types"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;
